subsys/bluetooth/audio/ascs.c

```diff
@@ -3128,17 +3227,53 @@ int bt_ascs_init(const struct bt_bap_unicast_server_cb *cb)
 
 void bt_ascs_cleanup(void)
 {
-	for (size_t i = 0; i < ARRAY_SIZE(ase_pool); i++) {
-		struct bt_ascs_ase *ase = &ase_pool[i];
+	for (size_t i = 0; i < ARRAY_SIZE(ascs.ase_pool); i++) {
+		struct bt_ascs_ase *ase = &ascs.ase_pool[i];
 
 		if (ase->conn != NULL) {
 			bt_ascs_release_ase(&ase->ep);
 		}
 	}
-
 	if (unicast_server_cb != NULL) {
 		bt_iso_server_unregister(&iso_server);
 		unicast_server_cb = NULL;
 	}
 }
+
+int bt_ascs_unregister(void)
+{
+	int err;
+	bool all_ases_idle = false;
+
+	bt_ascs_cleanup();
+	while (!all_ases_idle) {
+		all_ases_idle = true;
+		for (size_t i = 0; i < ARRAY_SIZE(ascs.ase_pool); i++) {
+			if (ascs.ase_pool[i].ep.status.state != BT_BAP_EP_STATE_IDLE) {
+/*
+*/
+				all_ases_idle = false;
+				k_sleep(K_MSEC(1));
+				break;
+			}
+		}
+	}
```

Comment:
What's preventing this from running infinitely? 

It can only break if some other thread does something. How is that guaranteed? 

I would suggest to just return an error in case any ASEs is non-idle, and let the application determine how/when to try again

Reply:
At this point the callback has been unregistered, this happens at the same time as the ASEs are ordered to be released. So, unless I'm mistaken and they can be affected from somewhere I don't understand, what you suggest can only happen if:

a) From another thread callback is registered again, during unregistration - between 'bt_ascs_cleanup' and this loop.
b) Usage of public API function 'bt_bap_unicast_server_config_ase' which calls 'bt_ascs_config_ase'.

> It can only break if some other thread does something. How is that guaranteed?

Based on what I wrote above, I concluded that such a thread can only be application thread, and I was unsure whether we actually support thread-safe functions everywhere in LE Audio. 

Yours is a good point though, and trivial change, and Ill add it!